File: backend/schemas/agents.py
# Copyright (c) 2026 Mechemsi. All rights reserved.
# Licensed under AGPLv3. See LICENSE file.
# Commercial licensing: info@mechemsi.com

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class AgentSettingsIn(BaseModel):
    display_name: str | None = None
    description: str | None = None
    supports_session: bool | None = None
    default_timeout: int | None = None
    max_retries: int | None = None
    environment_vars: dict | None = None
    cli_flags: dict | None = None
    command_templates: dict | None = None
    enabled: bool | None = None
    agent_type: str | None = None
    install_cmd: str | None = None
    post_install_cmd: str | None = None
    check_cmd: str | None = None
    prereq_check: str | None = None
    prereq_name: str | None = None
    needs_non_root: bool | None = None
    consolidated_default: bool | None = None
    agent_creates_pr: bool | None = None


class AgentSettingsOut(BaseModel):
    id: int
    agent_name: str
    display_name: str
    description: str
    supports_session: bool
    default_timeout: int
    max_retries: int
    environment_vars: dict
    cli_flags: dict
    command_templates: dict
    enabled: bool
    agent_type: str = "cli_binary"
    install_cmd: str | None = None
    post_install_cmd: str | None = None
    check_cmd: str | None = None
    prereq_check: str | None = None
    prereq_name: str | None = None
    needs_non_root: bool = False
    consolidated_default: bool = True
    agent_creates_pr: bool = False
    created_at: datetime | None = None
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class AgentInstallStatus(BaseModel):
    agent_name: str
    display_name: str
    description: str
    agent_type: str  # cli_binary | api_service
    installed: bool
    version: str | None = None
    path: str | None = None
    authenticated: bool | None = None
    auth_email: str | None = None
    auth_method: str | None = None


class AgentInstallRequest(BaseModel):
    agent_name: str


class AgentInstallResult(BaseModel):
    success: bool
    agent_name: str
    message: str | None = None
    error: str | None = None
    output: str | None = None


class UserAgentStatus(BaseModel):
    user: str
    agents: list[AgentInstallStatus]


class AgentManagementStatus(BaseModel):
    agents: list[AgentInstallStatus]
    by_user: list[UserAgentStatus] = []


class AgentInvocationOut(BaseModel):
    id: int
    run_id: int
    phase_execution_id: int | None = None
    workspace_server_id: int | None = None
    agent_name: str
    phase_name: str | None = None
    subtask_index: int | None = None
    subtask_title: str | None = None
    prompt_chars: int = 0
    response_chars: int = 0
    exit_code: int | None = None
    files_changed: list[str] | None = None
    duration_seconds: float | None = None
    estimated_tokens_in: int | None = None
    estimated_tokens_out: int | None = None
    estimated_cost_usd: float | None = None
    status: str = "running"
    error_message: str | None = None
    session_id: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    metadata_: dict[str, Any] | None = None
    # Don't include full prompt/response text in list views — too large

    model_config = ConfigDict(from_attributes=True)


class AgentInvocationDetail(AgentInvocationOut):
    """Includes full prompt and response text for detail view."""

    prompt_text: str | None = None
    response_text: str | None = None
    system_prompt_text: str | None = None

File: frontend/src/types/agents.ts
// Copyright (c) 2026 Mechemsi. All rights reserved.
// Licensed under AGPLv3. See LICENSE file.
// Commercial licensing: info@mechemsi.com

// Hard-coded allowlist of agents shown in pickers (chat session, step
// editor, agent install, etc.). The backend may still know about more
// agents in the DB — this is the user-facing surface only. To re-show
// an agent, add it here and make sure its AgentSettings row exists.
export const AGENT_NAMES = [
  "claude",
  "codex",
  "opencode",
] as const;
export type AgentName = (typeof AGENT_NAMES)[number];

/** Set form of AGENT_NAMES for ``.has()`` checks when filtering API responses. */
export const VISIBLE_AGENTS: ReadonlySet<string> = new Set(AGENT_NAMES);

export interface AgentSettings {
  id: number;
  agent_name: string;
  display_name: string;
  description: string;
  supports_session: boolean;
  default_timeout: number;
  max_retries: number;
  environment_vars: Record<string, string>;
  cli_flags: Record<string, string | boolean>;
  command_templates: Record<string, string | boolean>;
  enabled: boolean;
  agent_type: string;
  install_cmd: string | null;
  post_install_cmd: string | null;
  check_cmd: string | null;
  prereq_check: string | null;
  prereq_name: string | null;
  needs_non_root: boolean;
  consolidated_default: boolean;
  agent_creates_pr: boolean;
  created_at: string | null;
  updated_at: string | null;
}

export interface AgentInvocation {
  id: number;
  run_id: number;
  phase_execution_id: number | null;
  workspace_server_id: number | null;
  agent_name: string;
  phase_name: string | null;
  subtask_index: number | null;
  subtask_title: string | null;
  prompt_chars: number;
  response_chars: number;
  exit_code: number | null;
  files_changed: string[] | null;
  duration_seconds: number | null;
  estimated_tokens_in: number | null;
  estimated_tokens_out: number | null;
  estimated_cost_usd: number | null;
  status: string;
  error_message: string | null;
  session_id: string | null;
  started_at: string | null;
  completed_at: string | null;
  metadata_: Record<string, unknown> | null;
}

export interface AgentInvocationDetail extends AgentInvocation {
  prompt_text: string | null;
  response_text: string | null;
  system_prompt_text: string | null;
}

export interface AgentInstallStatus {
  agent_name: string;
  display_name: string;
  description: string;
  agent_type: string;
  installed: boolean;
  version: string | null;
  path: string | null;
  authenticated: boolean | null;
  auth_email: string | null;
  auth_method: string | null;
}

export interface UserAgentStatus {
  user: string;
  agents: AgentInstallStatus[];
}

export interface AgentManagementStatus {
  agents: AgentInstallStatus[];
  by_user: UserAgentStatus[];
}

export interface AgentInstallResult {
  success: boolean;
  agent_name: string;
  message: string | null;
  error: string | null;
  output: string | null;
}